### PR TITLE
Feature/velocity solvers

### PR DESCRIPTION
**Pull request by jbrnlz**
Checked line by line the different velocity solvers against their version in Ufe: SIA, SSA, SIA/SSA, and added a slightly different option for combining the SIA and SSA in the hybrid one: SIA velocity contribution is increasingly reduced as the SSA velocities get faster. This helps prevent instabilities in areas with strong surface gradients that are already sliding. Skipped DIVA on purpose, as some more checks are needed for that one in Ufe before a full port.

### CODE REVIEW

File: src/data_types_module.f90 [hunk 1]
```diff
@@ -63,13 +63,15 @@ module data_types_module
     REAL(dp), DIMENSION(:    ), allocatable :: v_surf_a
     REAL(dp), DIMENSION(:    ), allocatable :: u_surf_b
     REAL(dp), DIMENSION(:    ), allocatable :: v_surf_b
+    REAL(dp), DIMENSION(:    ), allocatable :: w_surf_a
```
Comment (v1kko) [inline src/data_types_module.f90:66]: no w_surf_b?

File: src/general_ice_model_data_module.f90 [hunk 2]
```diff
@@ -316,41 +319,44 @@ SUBROUTINE determine_grounded_fractions_a( mesh, ice)
...
+
+    do vi = mesh%vi1, mesh%vi2
 
       ! Skip border vertices
-      IF (mesh%edge_index( vi) > 0) THEN
-        ice%f_grnd_a( vi) = 0._dp
-        CYCLE
-      END IF
+      if (mesh%edge_index( vi) > 0) then
+        ice%f_grnd_a( vi) = real(ice%mask_land_a( vi), dp)
```
Comment (v1kko) [inline src/general_ice_model_data_module.f90:332]: A mask is either zero or one, is that what f_grnd_a should be ? (1.0, or 0.0)

File: src/ice_velocity_module.f90 [hunk 3]
```diff
@@ -252,28 +273,40 @@ subroutine solve_SSA(  mesh, ice)
...
-      !              ', u = [', MINVAL(ice%u_base_SSA_b), &
-      !              ' - ', MAXVAL(ice%u_base_SSA_b), ']'
+      ! Check for convergence or too many iterations
+      if (resid_UV < C%DIVA_visc_it_norm_dUV_tol .or. &
+          viscosity_iteration_i >= C%DIVA_visc_it_nit) then
+        has_converged = .true.
+      endif
+
+      ! Message printing: idealised case
+      if (par%master .and. C%choice_refgeo_init_ANT == 'idealised' .and. C%choice_refgeo_init_idealised == 'SSA_icestream') then
```
Comment (v1kko) [inline src/ice_velocity_module.f90:283]: Should this message printing stay? Or is it more like a debug statement.

**Review by v1kko — COMMENTED** [review]
I haven't fulliy checked the ice_velocity_module, because the github diff is not readable, however I trust that its good.

The rest looks good as well, some minor comments, which are more questions than things that have to be changed

File: src/general_ice_model_data_module.f90 (hunk 2 above)
Comment (jbrnlz) [reply src/general_ice_model_data_module.f90:332]: In this case, yes. For border edges, we assume they are fully grounded (1.0) or fully floating (0.0).

File: src/data_types_module.f90 (hunk 1 above)
Comment (jbrnlz) [reply src/data_types_module.f90:66]: Separate surf and base velocity components are purely diagnostic, and not needed in any operation involving the b- or c-grid right now.

File: src/data_types_module.f90 [hunk 4]
```diff
@@ -63,13 +63,15 @@ module data_types_module
...
     REAL(dp), DIMENSION(:    ), allocatable :: uabs_surf_a
     REAL(dp), DIMENSION(:    ), allocatable :: uabs_surf_b
 
     REAL(dp), DIMENSION(:    ), allocatable :: u_base_a                    ! Ice velocity at the base [m yr^-1]
     REAL(dp), DIMENSION(:    ), allocatable :: v_base_a
     REAL(dp), DIMENSION(:    ), allocatable :: u_base_b
     REAL(dp), DIMENSION(:    ), allocatable :: v_base_b
+    REAL(dp), DIMENSION(:    ), allocatable :: w_base_a
```
[the remark this answers is not in the document]
Comment (jbrnlz) [reply src/data_types_module.f90:74]: Separate surf and base velocity components are purely diagnostic, and not needed in any operation involving the b- or c-grid right now.

File: src/ice_velocity_module.f90 (hunk 3 above)
Comment (jbrnlz) [reply src/ice_velocity_module.f90:283]: Yeah, that message is printed only for that specific 'SSA_icestream' experiment, which is not often run. When run, however, you want almost always to have a look at the printed stats.

**Pull request merged**